pkg/flowaggregator/s3uploader/s3uploader.go

```diff
@@ -0,0 +1,453 @@
+	go func() {
+		defer p.exportWg.Done()
+		p.flowRecordPeriodicCommit()
+	}()
+}
+
+func (p *S3UploadProcess) stopExportProcess(flushQueue bool) {
+	p.mutex.Lock()
+	defer p.mutex.Unlock()
+	if !p.exportProcessRunning {
+		return
+	}
+	p.exportProcessRunning = false
+	defer p.uploadTicker.Stop()
+	p.stopCh <- stopPayload{
+		flushQueue: flushQueue,
+	}
+	p.exportWg.Wait()
+}
+
+func (p *S3UploadProcess) flowRecordPeriodicCommit() {
+	klog.InfoS("Starting S3 exporting process")
+	ctx := context.Background()
+	for {
+		select {
+		case stop := <-p.stopCh:
+			klog.InfoS("Stopping S3 exporting process")
+			if !stop.flushQueue {
+				return
+			}
+			ctx, cancelFn := context.WithTimeout(ctx, bufferFlushTimeout)
+			defer cancelFn()
+			err := p.batchUploadAll(ctx)
+			if err != nil {
+				klog.ErrorS(err, "Error when doing batchUploadAll on stop")
+			}
+			return
+		case <-p.uploadTicker.C:
+			err := p.batchUploadAll(ctx)
+			if err != nil {
+				klog.ErrorS(err, "Error when doing batchUploadAll on triggered timer")
+			}
+		}
+	}
+}
+
+// batchUploadAll uploads all buffers cached in bufferQueue and previous fail-
+// to-upload buffers stored in buffersToUpload. Returns error if encountered.
+func (p *S3UploadProcess) batchUploadAll(ctx context.Context) error {
+	p.queueMutex.Lock()
+	if p.cachedRecord == 0 && len(p.bufferQueue) == 0 {
+		p.queueMutex.Unlock()
+		return nil
+	}
+	if p.cachedRecord != 0 {
+		p.appendBufferToQueue()
+	}
+	// dump cached buffers from bufferQueue to buffersToUpload
+	for _, buf := range p.bufferQueue {
+		p.buffersToUpload = append(p.buffersToUpload, buf)
+		if len(p.buffersToUpload) > maxKeptBufferNum {
+			p.buffersToUpload = p.buffersToUpload[1:]
+		}
+	}
+	p.bufferQueue = p.bufferQueue[:0]
+	p.queueMutex.Unlock()
+
+	uploaded := 0
+	for _, buf := range p.buffersToUpload {
+		reader := bytes.NewReader(buf.Bytes())
+		err := p.uploadFile(reader, ctx)
+		if err != nil {
+			p.bufferQueue = p.bufferQueue[uploaded:]
+			return err
+		}
+		uploaded += 1
+	}
+	p.buffersToUpload = p.buffersToUpload[:0]
+	return nil
+}
+
+func (p *S3UploadProcess) writeRecordToBuffer(record *flowrecord.FlowRecord) {
+	var writer io.Writer
+	writer = p.currentBuffer
+	if *p.compress {
+		writer = p.gzipWriter
+	}
+	writeRecord(writer, record)
+	io.WriteString(writer, "\n")
+	p.cachedRecord += 1
+}
+
+func (p *S3UploadProcess) uploadFile(reader *bytes.Reader, ctx context.Context) error {
+	fileName := fmt.Sprintf("records-%s.csv", randSeq(12))
+	if *p.compress {
+		fileName += ".gz"
+	}
+	key := fileName
+	if p.bucketPrefix != "" {
+		key = fmt.Sprintf("%s/%s", p.bucketPrefix, fileName)
+	}
+	if _, err := p.s3UploaderAPI.Upload(ctx, &s3.PutObjectInput{
+		Bucket: aws.String(p.bucketName),
+		Key:    aws.String(key),
+		Body:   reader,
+	}, p.awsS3Uploader); err != nil {
+		return fmt.Errorf("error when uploading file to S3: %v", err)
+	}
+	return nil
+}
+
+// appendBufferToQueue appends currentBuffer to bufferQueue, and reset currentBuffer.
```

Comment:
add to this comment that the `queueMutex` must be held when calling this function